### PR TITLE
Added IngressClass objects support #117

### LINKED ISSUE #117: Add support for `IngressClass` objects
**What would you like to be added**:
Add support for reading and taking into account `IngressClass` objects.

**Why this is needed**:
If users do not specify `ingressClassName,` we need to understand the default `IngressClass` that should be used for the conversion.

### PR DESCRIPTION
**Pull request by Varun1300211**
**What type of PR is this?**
This pull request implements IngressClass support.

Add one of the following kinds:
/kind feature

**What this PR does / why we need it**:
It retrieves the default IngressClass based on its API group (networking.k8s.io) and a dedicated annotation (ingressclass.kubernetes.io/is-default-class).
If no explicit IngressClass is specified in the Ingress object or via annotation, the default class is used for resource processing.

**Which issue(s) this PR fixes**:

Fixes #117

**Does this PR introduce a user-facing change?**:

```NONE

``` 
Please do let me know if it was done in the correct way.

### CODE REVIEW

**Comment by Varun1300211** [comment]
Done all the changes requested by you @LiorLieberman, please let me know if anything else is required.

File: pkg/i2gw/providers/common/utils.go [hunk 1]
```diff
@@ -17,24 +17,50 @@ limitations under the License.
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/kubernetes"
 	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
 )
 
+var k8sClientset *kubernetes.Clientset
+
+func GetDefaultIngressClass() (string, error) {
+	ingressClasses, err := k8sClientset.NetworkingV1().IngressClasses().List(context.Background(), metav1.ListOptions{})
```
Comment (levikobi) [inline pkg/i2gw/providers/common/utils.go:35]: We support 2 modes for reading resources: from the cluster or a local file.
Here it seems that you're always reading from the cluster. We should come up with a way to take into account the user's choice

File: pkg/i2gw/providers/common/utils.go [hunk 2]
```diff
@@ -17,24 +17,50 @@ limitations under the License.
...
 	if ingress.Spec.IngressClassName != nil && *ingress.Spec.IngressClassName != "" {
 		ingressClass = *ingress.Spec.IngressClassName
 	} else if _, ok := ingress.Annotations[networkingv1beta1.AnnotationIngressClass]; ok {
 		ingressClass = ingress.Annotations[networkingv1beta1.AnnotationIngressClass]
 	} else {
-		ingressClass = ingress.Name
+		defaultIngressClass, err := GetDefaultIngressClass()
+		if err != nil {
+			return ""
+		}
```
Comment (levikobi) [inline pkg/i2gw/providers/common/utils.go:62]: We should probably return an error in this case

**Review by levikobi — COMMENTED** [review]
Thanks @Varun1300211 and welcome!
I left a few comments regarding some things that need to be taken into account.
If you have any questions or need any help feel free to let us know

**Comment by Varun1300211** [comment]
Hi @levikobi, I made the changes suggested by you. Please take a look.

File: pkg/i2gw/providers/common/utils.go [hunk 3]
```diff
@@ -17,24 +17,104 @@ limitations under the License.
...
+	var err error
+
+	switch extension {
+	case "yaml":
+		err = yaml.Unmarshal(data, ingressClasses)
+	case "json":
+		err = json.Unmarshal(data, ingressClasses)
+	default:
+		return nil, fmt.Errorf("unsupported file type: %s", extension)
+	}
```
Comment (levikobi) [inline pkg/i2gw/providers/common/utils.go:72]: Maybe I missed something but it seems like you're always sending an empty slice of data

File: pkg/i2gw/providers/common/utils.go [hunk 4]
```diff
@@ -17,24 +17,104 @@ limitations under the License.
...
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Fatal(errors.New("unable to create config"))
+	}
+
+	k8sClientset, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
```
Comment (levikobi) [inline pkg/i2gw/providers/common/utils.go:51]: This init function will always try to create a k8s client, even if the user doesn't have a cluster and want to convert some local file.

You'll need to pass the user's choice of where he wants to read resources from (cluster or file) and the k8s client, which was initialized in the i2gw package.

**Review by levikobi — CHANGES_REQUESTED** [review]
Thanks again @Varun1300211!
I think it may help you to try playing around with the tool for a bit and see the different options it provides. Afterwards maybe try to compile your version and see what changes/breaks.

This new addition needs to take into account different options the user may pass.

**Comment by LiorLieberman** [comment]
@Varun1300211 just checking in, do you need help with moving this forward?

**Comment by LiorLieberman** [comment]
@Varun1300211 Just checking in, please let us know if you need any help with moving this forward

**Comment by LiorLieberman** [comment]
@Varun1300211 a friendly ping, are you still planning to work on this?

**Pull request closed without merging**